### PR TITLE
feat: upload matrix ids

### LINKED ISSUE #1381: Always upload matrix_ids.json to GCS
**Author the user story for this feature**

As a user, I want to know all the matrix ids so I can investigate the test results.

**Describe the solution**

Always upload `matrix_ids.json` to gcs. Currently this file is generated and saved locally in `results/matrix_ids.json`

### PR DESCRIPTION
**Pull request by Sloox**
Fixes #1381 

1. Run any flank configuration which will not fail before executing test matrices.
2. Wait until execution finish.
3. Verify the matrices_ids.json is uploaded to the bucket.

### CODE REVIEW

File: test_runner/src/main/kotlin/ftl/gc/GcStorage.kt [hunk 1]
```diff
@@ -93,6 +95,17 @@ object GcStorage {
             logLn("Cannot upload performance metrics ${it.message}")
         }.getOrNull()
 
+    fun uploadMatricesId(args: IArgs, matrixMap: MatrixMap): String {
+        val matrixIdsPath: Path = if (args.useLocalResultDir())
+            Paths.get(args.localResultDir, FtlConstants.matrixIdsFile) else
+            Paths.get(args.localResultDir, matrixMap.runPath, FtlConstants.matrixIdsFile)
```
Comment (jan-gogo) [inline test_runner/src/main/kotlin/ftl/gc/GcStorage.kt:101]: The same logic exists in `updateMatrixFile`, IMO we can have a function for getting matrixIdsPath. For example `fun IArgs.matrixIdsPath(matrixMap: MatrixMap): Path`

Comment (Sloox) [reply test_runner/src/main/kotlin/ftl/gc/GcStorage.kt:101]: Sounds good!

File: test_runner/src/main/kotlin/ftl/gc/GcStorage.kt [hunk 2]
```diff
@@ -93,6 +95,15 @@ object GcStorage {
...
         }.getOrNull()
 
+    fun uploadMatricesId(args: IArgs, matrixMap: MatrixMap): String {
+        val matrixIdsPath = args.getMatrixFilePath(matrixMap)
+        return upload(
+            file = matrixIdsPath.toString(),
+            rootGcsBucket = args.resultsBucket,
+            runGcsPath = args.resultsDir
+        )
+    }
```
Comment (jan-gogo) [inline test_runner/src/main/kotlin/ftl/gc/GcStorage.kt:105]:
```suggestion
    fun uploadMatricesId(args: IArgs, matrixMap: MatrixMap): String =
        upload(
            file = args.getMatrixFilePath(matrixMap).toString(),
            rootGcsBucket = args.resultsBucket,
            runGcsPath = args.resultsDir
        )
```

**Comment by pawelpasterz** [comment]
Please, remember to update the integration tests

**Comment by adamfilipow92** [comment]
I know we have ```always``` in the ticket title but, what when we set ```disable-results-upload```? In this case, we not uploading any report or ```_shards.json``` maybe this file shouldn't be uploaded too?

**Comment by pawelpasterz** [comment]
> I know we have `always` in the ticket title but, what when we set `disable-results-upload`? In this case, we not uploading any report or `_shards.json` maybe this file shouldn't be uploaded too?

Actually good point!

**Comment by jan-gogo** [comment]
> I know we have `always` in the ticket title but, what when we set `disable-results-upload`? In this case, we not uploading any report or `_shards.json` maybe this file shouldn't be uploaded too?

I have added this to my suggestion https://github.com/Flank/flank/pull/1393#discussion_r540308399

**Comment by Sloox** [comment]
@pawelpasterz can you please verify the integration test yourside. to see if anything fails. I cannot even run them at all on 3 different machines (Windows/ Mac or ubuntu)

File: integration_tests/src/test/resources/compare/MultipleApksIT-compare [hunk 3]
```diff
@@ -1 +1 @@
...
  Uploading matrix_ids.json \.*
FetchArtifacts
  \.*
  Updating matrix file
\s*
Matrices webLink
  matrix-[a-zA-Z0-9]* https:\/\/console.firebase.google.com\/project\/flank-open-source\/testlab\/histories\/[.a-zA-Z0-9_-]*\/matrices\/[.a-zA-Z0-9_-]*(\/executions\/[.a-zA-Z0-9_-]*)?
  matrix-[a-zA-Z0-9]* https:\/\/console.firebase.google.com\/project\/flank-open-source\/testlab\/histories\/[.a-zA-Z0-9_-]*\/matrices\/[.a-zA-Z0-9_-]*(\/executions\/[.a-zA-Z0-9_-]*)?
  matrix-[a-zA-Z0-9]* https:\/\/console.firebase.google.com\/project\/flank-open-source\/testlab\/histories\/[.a-zA-Z0-9_-]*\/matrices\/[.a-zA-Z0-9_-]*(\/executions\/[.a-zA-Z0-9_-]*)?
  matrix-[a-zA-Z0-9]* https:\/\/console.firebase.google.com\/project\/flank-open-source\/testlab\/histories\/[.a-zA-Z0-9_-]*\/matrices\/[.a-zA-Z0-9_-]*(\/executions\/[.a-zA-Z0-9_-]*)?
```
Comment (Sloox) [inline integration_tests/src/test/resources/compare/MultipleApksIT-compare:1]: @adamfilipow92 thanks. 👍

Comment (adamfilipow92) [reply integration_tests/src/test/resources/compare/MultipleApksIT-compare:1]: No problem 😄

**Review by adamfilipow92 — APPROVED** [review]
- [X] Code looks good
- [X] IT updated
Great work! 👍

File: test_runner/src/main/kotlin/ftl/gc/GcStorage.kt [hunk 4]
```diff
@@ -93,6 +95,15 @@ object GcStorage {
...
         }.getOrNull()
 
+    fun uploadMatricesId(args: IArgs, matrixMap: MatrixMap) {
+        if (args.disableResultsUpload) return
+        upload(
+            file = args.getMatrixFilePath(matrixMap).toString(),
+            rootGcsBucket = args.resultsBucket,
+            runGcsPath = args.resultsDir
+        )
+    }
```
Comment (piotradamczyk5) [inline test_runner/src/main/kotlin/ftl/gc/GcStorage.kt:105]: could we add unit test to it?

Comment (Sloox) [reply test_runner/src/main/kotlin/ftl/gc/GcStorage.kt:105]: Actually @jan-gogo and I were chatting about it. We decided against it.

Comment (jan-gogo) [reply test_runner/src/main/kotlin/ftl/gc/GcStorage.kt:105]: yep, but still, we can decide to add if someone decides that unit test for it is important.

**Review by jan-gogo — COMMENTED** [review]
The implementation looks well, but I need to verify if a matrix_ids.json is uploaded to the `bucket`. Currently, Google servers are down.

**Comment by jan-gogo** [comment]
>     * [x]  Code looks good
> 
>     * [x]  IT updated
>       Great work! 👍

Next time attach links to the results so anyone could verify them.

**Review by jan-gogo — APPROVED** [review]
looks ok:
* https://console.cloud.google.com/storage/browser/test-lab-v9cn46bb990nx-kz69ymd4nm9aq/2020-12-11_13-59-26.959002_prjU;tab=objects?prefix=&forceOnObjectsSortingFiltering=false
* https://console.cloud.google.com/storage/browser/test-lab-v9cn46bb990nx-kz69ymd4nm9aq/2020-12-11_13-57-45.704878_lOlY;tab=objects?pageState=(&prefix=&forceOnObjectsSortingFiltering=false

**Pull request merged**